src/stan/math/prim/mat/meta/index_type.hpp

```diff
@@ -21,6 +21,16 @@ namespace stan {
       typedef typename Eigen::Matrix<T,R,C>::Index type;
     };
 
+    template <typename T>
+    struct index_type<Eigen::MatrixBase<T> > {
+      typedef typename index_type<T>::type type;
+    };
+
+    template <typename T>
+    struct index_type<Eigen::Map<T> > {
```

Comment:
Yes, `Eigen::Map` signals to Eigen that the memory allocated for the matrix is not controlled by Eigen whereas `Eigen::MatrixBase` is a generic wrapper class that should wrap any Eigen dense matrix where the actual matrix type is `T`. So they do different things.